server/util/networking/networking.go

```diff
@@ -213,6 +226,136 @@ func RemoveNetNamespace(ctx context.Context, netNamespace string) error {
+// across container instances. This pooling helps to reduce the performance
+// overhead associated with rapidly creating and destroying networks along with
+// all of their associated configuration.
+//
+// TODO: consolidate logic so that VM networks can be pooled too. VMs have an
+// additional TAP device which isn't needed for container networks, but most of
+// the other setup is the same.
+type ContainerNetworkPool struct {
+	sizeLimit int
+
+	mu           sync.Mutex
+	resources    []*ContainerNetwork
+	shuttingDown bool
+}
+
+func NewContainerNetworkPool(sizeLimit int) *ContainerNetworkPool {
+	if sizeLimit < 0 {
+		sizeLimit = defaultContainerNetworkPoolSizeLimit
+	}
+	return &ContainerNetworkPool{
+		sizeLimit: sizeLimit,
+	}
+}
+
+// Get returns a pooled veth pair, or nil if there are no pooled veth pairs
+// available.
+func (p *ContainerNetworkPool) Get(ctx context.Context) (n *ContainerNetwork) {
+	n = p.get()
+	if n == nil {
+		return nil
+	}
+
+	// If we fail to fully set up the network, then we're on the hook for
+	// cleaning it up, since we already took it from the pool.
+	defer func() {
+		if n != nil {
+			return
+		}
+		ctx, cancel := background.ExtendContextForFinalization(ctx, networkingCleanupTimeout)
+		defer cancel()
+		if err := n.Cleanup(ctx); err != nil {
+			log.CtxErrorf(ctx, "Failed to clean up pooled network in partially set up state: %s", err)
+		}
+	}()
+
+	// Assign a new IP before returning the network from the pool.
+	network, err := hostNetAllocator.Get()
+	if err != nil {
+		log.CtxErrorf(ctx, "Failed to allocate new IP range for pooled network: %s", err)
+		return nil
+	}
+	n.vethPair.network = network
+
+	// Assign IPs to the host and namespaced side, and create the default route
+	// in the namespace.
+	if err := attachAddressToVeth(ctx, "" /*no namespace*/, network.HostIPWithCIDR(), n.vethPair.hostDevice); err != nil {
+		log.CtxErrorf(ctx, "Failed to attach address to pooled host veth interface: %s", err)
+		return nil
+	}
+	if err := attachAddressToVeth(ctx, n.vethPair.netNamespace, network.NamespacedIPWithCIDR(), n.vethPair.namespacedDevice); err != nil {
+		log.CtxErrorf(ctx, "Failed to attach address to pooled namespaced veth interface: %s", err)
+		return nil
+	}
+	if err := runCommand(ctx, namespace(n.vethPair.netNamespace, "ip", "route", "add", "default", "via", network.HostIP())...); err != nil {
+		log.CtxErrorf(ctx, "Failed to set up default route in namespace: %s", err)
+		return nil
+	}
+
+	return n
+}
+
+func (p *ContainerNetworkPool) get() *ContainerNetwork {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	if len(p.resources) == 0 {
+		return nil
+	}
+
+	head, tail := p.resources[0], p.resources[1:]
+	p.resources = tail
+	return head
+}
+
+// Add adds a veth pair to the pool.
+// It returns whether the veth pair was successfully added.
+// The caller should clean up the veth pair if this returns false.
+func (p *ContainerNetworkPool) Add(ctx context.Context, n *ContainerNetwork) (ok bool) {
```

Comment:
nit: don't need a named return var

Reply:
When a func returns a non-obvious value such as `bool` I think using a named value helps for documentation purposes.